cli/pcluster/utils.py

```diff
@@ -199,3 +202,135 @@ def get_supported_schedulers():
+    aws_reserved_ip = 6
+    min_bitmask_length = 28
+    target_bitmask_length = min(
+        32 - ((next_power_of_2(target_size + aws_reserved_ip) - 1).bit_length()), min_bitmask_length
+    )
+    subnet_size = 2 ** (32 - target_bitmask_length)
+    vpc_begin_address_decimal, vpc_end_address_decimal = _get_cidr_limits_as_decimal(vpc_cidr)
+
+    if vpc_end_address_decimal - vpc_begin_address_decimal + 1 < subnet_size:  # if we do not have enough space
+        return None
+
+    if not occupied_cidr:  # if we have space and no occupied cidr
+        return _decimal_ip_limits_to_cidr(vpc_begin_address_decimal, vpc_begin_address_decimal + subnet_size)
+
+    occupied_cidr_max_bitmask = max([int(subnet_cidr.split("/")[1]) for subnet_cidr in occupied_cidr])
+    if occupied_cidr_max_bitmask > target_bitmask_length:
+        # This means that it's smaller, so we need to make it bigger
+        occupied_cidr = _expand_cidrs(occupied_cidr, min_size=target_bitmask_length)
+
+    # subnets_number is a list of pair(begin ip, end ip) obtained from the cidr. So for example
+    # 10.0.0.0/17 = 10.0.0.0, 10.0.127.255
+    begin_ip_index = 0
+    end_ip_index = 1
+    subnets_limits = [_get_cidr_limits_as_decimal(subnet) for subnet in occupied_cidr]
+    subnets_limits.sort(key=lambda x: x[1])  # sort by ending numbers, sorting by beginning is the same
+    # to check for space between the last occupied and the end of the vpc
+    subnets_limits.append((vpc_end_address_decimal, vpc_end_address_decimal))
+
+    if (subnets_limits[0][begin_ip_index] - vpc_begin_address_decimal) > subnet_size:
+        return _decimal_ip_limits_to_cidr(vpc_begin_address_decimal, vpc_begin_address_decimal + subnet_size)
+
+    #  Looking at space between occupied cidrs
+    for index in range(1, len(subnets_limits)):
+        begin_number = subnets_limits[index][begin_ip_index]
+        end_previous_number = subnets_limits[index - 1][end_ip_index]
+        if begin_number - end_previous_number > subnet_size:
+            return _decimal_ip_limits_to_cidr(end_previous_number + 1, end_previous_number + subnet_size)
+    return None
+
+
+def _decimal_ip_limits_to_cidr(begin, end):
+    """Given begin and end ip (as decimals number), return the CIDR that begins with begin ip and ends with end ip."""
+    return str(
+        next(
+            summarize_address_range(
+                ip_address(unicode(socket.inet_ntoa(struct.pack("!L", begin)))),
+                ip_address(unicode(socket.inet_ntoa(struct.pack("!L", end)))),
+            )
+        )
+    )
+
+
+def _get_cidr_limits_as_decimal(cidr):
+    """Given a cidr, return the begin ip and the end ip as decimal."""
+    address = ip_network(unicode(cidr))
+    return _ip_to_decimal(str(address[0])), _ip_to_decimal(str(address[-1]))
+
+
+def _ip_to_decimal(ip):
+    """Transform an ip into its decimal representantion."""
+    return int(bin(struct.unpack("!I", socket.inet_aton(ip))[0]), 2)
```

Comment:
Why don't use: `int(ipaddress.IPv4Address(ip))` ?
https://docs.python.org/3/library/ipaddress.html#conversion-to-strings-and-integers